src/test/msgr/test_async_networkstack.cc

```diff
@@ -0,0 +1,1050 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2016 XSky <haomai@xsky.com>
+ *
+ * Author: Haomai Wang <haomaiwang@gmail.com>
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#include <algorithm>
+#include <atomic>
+#include <iostream>
+#include <string>
+#include <set>
+#include <vector>
+#include <gtest/gtest.h>
+
+#include "acconfig.h"
+#include "include/Context.h"
+#include "common/ceph_argparse.h"
+#include "global/global_init.h"
+
+#include "msg/async/Event.h"
+#include "msg/async/Stack.h"
+
+#if GTEST_HAS_PARAM_TEST
+
+class NetworkWorkerTest : public ::testing::TestWithParam<const char*> {
+ public:
+  std::shared_ptr<NetworkStack> stack;
+  string addr, port_addr;
+
+  NetworkWorkerTest() {}
+  virtual void SetUp() {
+    cerr << __func__ << " start set up " << GetParam() << std::endl;
+    addr = "127.0.0.1:15000";
+    port_addr = "127.0.0.1:15001";
+    stack = NetworkStack::create(g_ceph_context, GetParam());
+    stack->start();
+  }
+  virtual void TearDown() {
+    stack->stop();
+  }
+  string get_addr() const {
+    return addr;
+  }
+  string get_ip_different_port() const {
+    return port_addr;
+  }
+  string get_different_ip() const {
+    return "10.0.123.100:4323";
+  }
+  EventCenter *get_center(unsigned i) {
+    return &stack->get_worker(i)->center;
+  }
+  Worker *get_worker(unsigned i) {
+    return stack->get_worker(i);
+  }
+  template<typename func>
+  class C_dispatch : public EventCallback {
+    Worker *worker;
+    func f;
+    bool done;
```

Comment:
might want to use `atomic<bool>` here instead. you will have a racing condition here.